core/chains/evm/client/simulated_backend_client.go

```diff
@@ -461,7 +461,9 @@ func (c *SimulatedBackendClient) BatchCallContext(ctx context.Context, b []rpc.B
 			b[i].Error = err
 		case "eth_getBlockByNumber":
 			if _, ok := elem.Result.(*evmtypes.Head); !ok {
-				return errors.Errorf("SimulatedBackendClient expected return type of *evmtypes.Head for eth_getBlockByNumber, got type %T", elem.Result)
+				if _, ok = elem.Result.(*evmtypes.Block); !ok {
+					return errors.Errorf("SimulatedBackendClient expected return type of [*evmtypes.Head] or [*evmtypes.Block] for eth_getBlockByNumber, got type %T", elem.Result)
+				}
```

Comment:
Is this sufficient? Down on 490, an explicit `*evmtypes.Head` is still used, so I would have thought we'd need to match by alternatively return an `*evmtypes.Block` :thinking:

Or, instead of enumerating all the types that we expect, we could fix it once in a forward compatible way by constructing a JSON string to unmarshal in to `elem.Result` directly.

Reply:
So evmtypes.Head is a light block, without full Txs.
evmtypes.Block is a block with full Txs in it.

Only BlockHistoryEstimator is using evmtypes.Block, to fetch full blocks and read Tx details.
However, there's no actual test case currently which uses SimulatedBackendClient, and expects it to return full Block with Txs.
If you see lines 479-481, it explicitly says this client doesn't support full Txs.

Trying to fix it generically is more work, plus this function will need to inject custom Txs, etc. All of that is useful only if there's a test-case that needs it. Currently there's none.

So I just fixed it enough that it doesn't break any existing tests. When someone adds more advanced test cases, then this will need updating.

Does that sounds reasonable?

Reply:
Also, it would be valid to use a complete block type (with full txes) for the result of a request for only the header (without txes). There is just a bool parameter, and the type is not enforced.